test/main_test.cpp

```diff
@@ -0,0 +1,218 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#include <iostream>
+#include "lest/lest.hpp"
+#include "frame.h"
+#include "flatbuffers.h"
+
+namespace torchcraft {
+namespace replayer {
+
+// For comparing properties on test objects
+#define E(property) && a.property == b.property
+
+bool operator==(const Bullet& a, const Bullet& b) {
+  return true E(type) E(x) E(y);
+}
+
+bool operator==(const Action& a, const Action& b) {
+  return true E(action) E(uid) E(aid);
+}
+
+bool operator==(const Resources& a, const Resources& b) {
+  return true
+    E(ore) E(gas) E(used_psi) E(total_psi)
+    E(upgrades) E(upgrades_level) E(techs);
+}
```

Comment:
All these structs should have default equality operators.

Reply:
Overloading operators is a bit of a maintenance trap, especially equality operators. It's really hard to search for usages of overloaded operators, so it's hard to discover all the consequences of modifying or deleting one.

Since these operators are only necessary for the test, I've limited their scope to the test. We didn't need them before and there doesn't seem to be an immediate indication that we'll need them again later.

Reply:
What I meant is that there should be equality operators present already and that there's no need to introduce custom ones here.

Reply:
Some types already had equality operators; I added them here for the ones that didn't.

Reply:
Via https://github.com/TorchCraft/TorchCraft/blob/develop/include/frame.h#L38 -- Order and UnitCommand have equality operators; nothing else does.

Reply:
Yes, but they only have them because certain fields should be ignored in the comparison. If you compare all fields like for `Resources`, you can just use the ones that are defined by default.

Reply:
Pardon my confusion. It sounds like you're suggesting that C++ has default equality operators for structs. Is that the case, or am I misunderstanding?

Reply:
Dang, it seems like I was the one being confused here! So yeah, there are no default equality operators, so what you're doing here is legit after all. Sorry!